File: LuaLib/hsgmParser.lua
-- Constants
local tileSize = 16
local width = 256
local height = 256

-- Function to parse the map file
local function parseMapFile(path)
    local file = io.open(path, "r")
    if not file then
        error("Could not open file: " .. path)
    end

    local textures = {}
    local colliders = {}
    local triggers = {}
    local binding = {}

    local section = nil

    for line in file:lines() do
        line = line:match("^%s*(.-)%s*$") -- Trim whitespace

        if line:match("^%[TEXTURES%]$") then
            section = "textures"
        elseif line:match("^%[COLLIDERS%]$") then
            section = "colliders"
        elseif line:match("^%[TRIGGERS%]$") then
            section = "triggers"
        elseif section == "textures" then
            local key, value = line:match("^(.-)%s*=%s*(.-)$")
            if key and value then
                local x, y = value:match("%[(%d+),%s*(%d+)%]")
                textures[key] = {tonumber(x), tonumber(y)}
            end
        elseif section == "colliders" then
            local key, value = line:match("^(.-)%s*=%s*(.-)$")
            if key and value then
                local x, y = value:match("%[(%d+),%s*(%d+)%]")
                colliders[key] = {tonumber(x), tonumber(y)}
            end
        elseif section == "triggers" then
            local key, value = line:match("^(.-)%s*=%s*(.-)$")
            if key and value then
                local x, y = value:match("%[(%d+),%s*(%d+)%]")
                triggers[key] = {tonumber(x), tonumber(y)}
            end
        end
    end

    file:close()

    return textures, colliders, triggers, binding
end

-- Basic Simulation of zlib compression
local function basicCompress(data)
    local compressed = {}
    for i = 1, #data do
        compressed[i] = data:byte(i) -- Very simple "compression" (not real compression)
    end
    return string.char(table.unpack(compressed))
end

-- PNG Helper Functions
local function crc32(str)
    local crc = 0xFFFFFFFF
    for i = 1, #str do
        local byte = str:byte(i)
        crc = crc ~ byte
        for _ = 0, 7 do
            if (crc & 1) ~= 0 then
                crc = (crc >> 1) ~ 0xEDB88320
            else
                crc = crc >> 1
            end
        end
    end
    return crc ~ 0xFFFFFFFF
end

local function createChunk(type, data)
    local length = #data
    local crc = crc32(type .. data)
    return string.pack(">I4", length) .. type .. data .. string.pack(">I4", crc)
end

local function createIHDRChunk()
    local widthBytes = string.pack(">I4", width)
    local heightBytes = string.pack(">I4", height)
    local ihdrData = widthBytes .. heightBytes .. string.char(8, 6, 0, 0, 0) -- 8-bit depth, RGB
    return createChunk("IHDR", ihdrData)
end

local function createIDATChunk(pixelData)
    -- Compress pixel data
    local compressedData = basicCompress(pixelData)
    return createChunk("IDAT", compressedData)
end

local function createIENDChunk()
    return createChunk("IEND", "")
end

local function writePng(path, pixelData)
    local file = io.open(path, "wb")
    if not file then
        error("Could not open file: " .. path)
    end

    file:write("\137PNG\r\n\26\n")
    file:write(createIHDRChunk())
    file:write(createIDATChunk(pixelData))
    file:write(createIENDChunk())
    file:close()
end

-- Image Generation
local function generatePixelData(items, binding, isCollider)
    local pixelData = {}

    for y = 0, height - 1 do
        for x = 0, width - 1 do
            local index = (y * width + x) * 4
            local color = {0, 0, 0, 255} -- Default color

            if isCollider then
                -- Get color from binding (placeholder logic)
                color = {0, 0, 255, 255}
            else
                -- Get color from binding (placeholder logic)
                color = {255, 0, 0, 255}
            end

            pixelData[index + 1] = color[1]
            pixelData[index + 2] = color[2]
            pixelData[index + 3] = color[3]
            pixelData[index + 4] = color[4]
        end
    end

    return pixelData
end

File: JavascriptLib/hsgmParser.js
const fs = require('fs');
const zlib = require('zlib'); // Node.js built-in zlib module for compression

// Constants
const tileSize = 16;
const width = 256;
const height = 256;

// Function to parse the map file
function parseMapFile(path) {
    const data = fs.readFileSync(path, 'utf-8');
    const textures = {};
    const colliders = {};
    const triggers = {};
    let section = null;

    data.split('\n').forEach(line => {
        line = line.trim();

        if (line === '[TEXTURES]') {
            section = 'textures';
        } else if (line === '[COLLIDERS]') {
            section = 'colliders';
        } else if (line === '[TRIGGERS]') {
            section = 'triggers';
        } else if (section === 'textures') {
            const [key, value] = line.split('=').map(part => part.trim());
            if (key && value) {
                const [x, y] = value.slice(1, -1).split(',').map(Number);
                textures[key] = [x, y];
            }
        } else if (section === 'colliders') {
            const [key, value] = line.split('=').map(part => part.trim());
            if (key && value) {
                const [x, y] = value.slice(1, -1).split(',').map(Number);
                colliders[key] = [x, y];
            }
        } else if (section === 'triggers') {
            const [key, value] = line.split('=').map(part => part.trim());
            if (key && value) {
                const [x, y] = value.slice(1, -1).split(',').map(Number);
                triggers[key] = [x, y];
            }
        }
    });

    return { textures, colliders, triggers };
}

// Function to calculate CRC32
function crc32(str) {
    const table = Array.from({ length: 256 }, (_, i) => {
        let crc = i;
        for (let j = 8; j > 0; j--) {
            if (crc & 1) {
                crc = (crc >>> 1) ^ 0xEDB88320;
            } else {
                crc >>>= 1;
            }
        }
        return crc;
    });

    let crc = 0xFFFFFFFF;
    for (let i = 0; i < str.length; i++) {
        const byte = str.charCodeAt(i) & 0xFF;
        crc = (crc >>> 8) ^ table[(crc ^ byte) & 0xFF];
    }

    return (crc ^ 0xFFFFFFFF) >>> 0;
}

// Function to create PNG chunks
function createChunk(type, data) {
    const length = data.length;
    const crc = crc32(type + data);
    const lengthBuf = Buffer.alloc(4);
    lengthBuf.writeUInt32BE(length, 0);
    const typeBuf = Buffer.from(type);
    const crcBuf = Buffer.alloc(4);
    crcBuf.writeUInt32BE(crc, 0);
    return Buffer.concat([lengthBuf, typeBuf, Buffer.from(data), crcBuf]);
}

// Function to create IHDR chunk
function createIHDRChunk() {
    const widthBuf = Buffer.alloc(4);
    widthBuf.writeUInt32BE(width, 0);
    const heightBuf = Buffer.alloc(4);
    heightBuf.writeUInt32BE(height, 0);
    const ihdrData = Buffer.concat([widthBuf, heightBuf, Buffer.from([8, 6, 0, 0, 0])]);
    return createChunk('IHDR', ihdrData.toString('binary'));
}

// Function to create IDAT chunk
function createIDATChunk(pixelData) {
    const compressedData = zlib.deflateSync(pixelData); // Using Node.js zlib for actual compression
    return createChunk('IDAT', compressedData.toString('binary'));
}

// Function to create IEND chunk
function createIENDChunk() {
    return createChunk('IEND', '');
}

// Function to write PNG file
function writePng(path, pixelData) {
    const file = fs.createWriteStream(path);
    file.write(Buffer.from([137, 80, 78, 71, 13, 10, 26, 10])); // PNG signature
    file.write(createIHDRChunk());
    file.write(createIDATChunk(pixelData));
    file.write(createIENDChunk());
    file.end();
}

// Function to generate pixel data
function generatePixelData(items, binding, isCollider) {
    const pixelData = Buffer.alloc(width * height * 4);

    for (let y = 0; y < height; y++) {
        for (let x = 0; x < width; x++) {
            const index = (y * width + x) * 4;
            const color = isCollider ? [0, 0, 255, 255] : [255, 0, 0, 255];
            pixelData[index] = color[0]; // R
            pixelData[index + 1] = color[1]; // G
            pixelData[index + 2] = color[2]; // B
            pixelData[index + 3] = color[3]; // A
        }
    }

    return pixelData;
}

